src/pages/index.js

```diff
@@ -1,81 +1,199 @@
 }
-// Section for rendering cards
-const cardSection = new Section(
-  {
-    items: constants.initialCards,
-    renderer: renderCard,
-  },
-  ".cards__list"
-);
-cardSection.renderItems();
-
-function renderCard(item) {
-  const cardElement = createCard(item);
-  cardSection.addItem(cardElement);
-}
-// UserInfo
+// UserInfo and Cards
 const userInformation = new UserInfo({
   name: ".profile__title",
   description: ".profile__description",
+  avatar: ".profile__image",
 });
 
+Promise.all([api.getUserInfo(), api.getInitialCards()])
+  // UserInfo
+  .then(([userData, cards]) => {
+    userInformation.setUserInfo({
+      name: userData.name,
+      description: userData.about,
+    });
+    userInformation.setAvatar(userData.avatar);
+    // Cards
+    cardSection = new Section(
+      {
+        items: cards,
+        renderer: (cardData) => {
+          const cardEl = createCard(cardData);
+          cardSection.addItem(cardEl);
+        },
+      },
+      ".cards__list"
+    );
+    cardSection.renderItems();
+  })
+  .catch((err) => {
+    console.error(err);
+  });
+// PopupWithImage and image preview
+const popupImage = new PopupWithImage("#modal-preview");
+popupImage.setEventListeners();
+
+function handleImageClick(data) {
+  popupImage.open(data);
+}
+// Likes
+function handleLikeClick(cardId) {
+  if (cardId._isLiked) {
+    api
+      .removeLike(cardId._id)
+      .then(() => {
+        cardId._renderLikes();
+        cardId._isLiked = false;
+      })
+      .catch((err) => {
+        console.error(err);
+      });
+  } else {
+    api
+      .addLike(cardId._id)
+      .then(() => {
+        cardId._renderLikes();
+        cardId._isLiked = true;
+      })
+      .catch((err) => {
+        console.error(err);
+      });
+  }
+}
 // PopupWithForm
-const addCardModal = new PopupWithForm("#add-modal", handleAddCardSubmit);
-addCardModal.setEventListeners();
+const avatarModal = new PopupWithForm("#avatar-modal", handleAvatarSubmit);
+avatarModal.setEventListeners();
 
 const profileEditModal = new PopupWithForm("#edit-modal", handleEditSubmit);
 profileEditModal.setEventListeners();
 
+const addCardModal = new PopupWithForm("#add-modal", handleAddCardSubmit);
+addCardModal.setEventListeners();
 // Handlers
-function handleImageClick(data) {
-  popupImage.open(data);
+function handleAvatarSubmit({ link }) {
+  avatarModal.renderLoad(true);
+  api
+    .updateAvatar(link)
+    .then((res) => {
+      userInformation.setAvatar(res.avatar);
+      avatarModal.close();
+      constants.avatarForm.reset();
+      avatarFormValidator.toggleBtnState();
+    })
+    .catch((err) => {
+      console.error("Oops...Failed to update avatar:", err);
+    })
+    .finally(() => {
+      avatarModal.renderLoad(false);
+    });
 }
-
 function handleEditSubmit(data) {
-  userInformation.setUserInfo({
-    title: data.title,
-    description: data.description,
-  });
-  profileEditModal.close();
+  profileEditModal.renderLoad(true);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}

```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```